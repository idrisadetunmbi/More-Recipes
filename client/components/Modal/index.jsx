import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import PropTypes from 'prop-types';

import SignIn from '../SignIn';
import SignUp from '../SignUp';
import CreateRecipe from '../CreateRecipe';
import './index.scss';

/**
 * @class Modal
 * @extends {React.Component}
 */
export default class Modal extends Component {
  /**
   * @returns {void}
   * @memberOf Modal
   */
  componentDidMount() {
    document.addEventListener('keydown', this.escKeyPress);
    document.body.style.overflow = 'hidden';
  }

  /**
   * @returns {void}
   * @memberOf Modal
   */
  componentWillUnmount() {
    document.body.style.overflow = '';
    document.removeEventListener('keydown', this.escKeyPress);
  }

  /**
   * @param {Object} event - DOM event object
   *
   * @returns {void}
   * @memberOf Modal
   */
  back = (event) => {
    event.stopPropagation();
    this.props.history.replace(this.props.location.state.previousLocation);
  };

  /**
   * @param {Object} event - DOM event object
   *
   * @returns {void}
   * @memberOf Modal
   */
  escKeyPress = (event) => {
    // if esc key is pressed
    if (event.keyCode === 27) {
      this.back(event);
    }
  };

  /**
   * @returns {JSX.Element} - Modal
   * @memberOf Modal
   */
  render() {
    return (
      <div onClick={this.back} id="modal-component">
        <div className="modal-overlay" />
        <div className="modal open">
          <div className="modal-content" onClick={event => event.stopPropagation()}>
            <div className="container">
              <div className="row">
                <div className="">
                  <Switch>
                    <Route path="/signin" component={SignIn} />
                    <Route path="/signup" component={SignUp} />
                    <Route path="/recipes/create" component={CreateRecipe} />
                  </Switch>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Modal.propTypes = {
  history: PropTypes.shape().isRequired,
  location: PropTypes.shape().isRequired,
};
