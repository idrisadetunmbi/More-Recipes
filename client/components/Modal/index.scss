#modal-component {
  .modal-overlay {
    z-index: 1002;
    display: block;
    opacity: 0.5;
  };
  .modal {
    z-index: 1003;
    display: block;
    opacity: 1;
    transform: scaleX(1);
    top: 10%;
    max-height: 100% !important;
    @media (min-width: 996px) {
      width: 40%;
    }
    // TODO: set more media queries for modal width
  }
}