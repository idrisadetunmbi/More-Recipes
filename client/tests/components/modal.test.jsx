import React from 'react';
import renderer from 'react-test-renderer';
import { StaticRouter as Router } from 'react-router-dom';
import { mount } from 'enzyme';

import Modal from '../../components/Modal';

const setUpComponent = (dynamicProps) => {
  const props = {
    ...dynamicProps,
    history: {
      replace: jest.fn(),
    },
    location: { state: {} },
  };
  const root = mount(<Router><Modal {...props} /></Router>);
  const component = root.find(Modal);
  const elements = {
    containerDiv: component.find('#modal-component'),
    modalContent: component.find('.modal-content'),
  };
  const methodSpy = name => jest.spyOn(component.instance(), name);

  const methods = {
    componentDidMount: methodSpy('componentDidMount'),
    back: methodSpy('back'),
    escKeyPress: methodSpy('escKeyPress'),
    componentWillUnmount: methodSpy('componentWillUnmount'),
  };
  root.instance().forceUpdate();
  return {
    root,
    component,
    elements,
    methods,
  };
};

describe('Modal component', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<Router><Modal /></Router>);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('calls back method when containerDiv is clicked ', () => {
    const { methods: { back }, elements: { containerDiv } } = setUpComponent();
    containerDiv.simulate('click');
    expect(back).toHaveBeenCalled();
  });

  it('', () => {
    const { elements: { modalContent } } = setUpComponent();
    modalContent.simulate('click');
  });

  it('calls component will unmount when component unmounts ', () => {
    const { methods: { componentWillUnmount }, root } = setUpComponent();
    root.unmount();
    expect(componentWillUnmount).toHaveBeenCalled();
  });
});
