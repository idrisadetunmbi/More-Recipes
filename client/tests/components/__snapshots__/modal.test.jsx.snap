// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component matches snapshot 1`] = `
<div
  id="modal-component"
  onClick={[Function]}
>
  <div
    className="modal-overlay"
  />
  <div
    className="modal open"
  >
    <div
      className="modal-content"
      onClick={[Function]}
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
